src/main/java/org/testng/DependencyMap.java

```diff
@@ -60,19 +58,38 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
     	return fromMethod;
     }
     for (ITestNGMethod m : l) {
-      // If they are in the same class hierarchy, they must belong to the same instance,
-      // otherwise, it's a method depending on a method in a different class so we
-      // don't bother checking the instance
-      if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass())) {
-        if (m.getInstance() == fromMethod.getInstance()) return m;
-      } else {
+      if (isSameInstance(fromMethod, m)) {
```

Comment:
Just merge the 2:
`if (isSameInstance(fromMethod, m) || belongToDifferentClassHierarchy(fromMethod, m) || hasNoInstance(fromMethod, m))`

Reply:
Yeah that looks more elegant and clean. Let me see if I can keep the tests happy with this change :)

Reply:
@juherr - I have changed it as per your recommendations. Please check